### ISSUE TITLE
Webhook: Invalid yaml and unstartable service with FOSS

### ISSUE THREAD

**Issue by rnelson0** [acidprime/r10k#162]
I am running into the issue described in #157, where select! causes an error generating the webhook yaml.
```
[root@puppet ~]# puppet agent -t --environment webhook
Info: Retrieving pluginfacts
Info: Retrieving plugin
Info: Loading facts
Error: Could not retrieve catalog from remote server: Error 400 on SERVER: Failed to parse template r10k/webhook.yaml.erb:
  Filepath: /etc/puppet/environments/webhook/modules/r10k/templates/webhook.yaml.erb
  Line: 1
  Detail: undefined method `select!' for #<Hash:0x7facf40a27d0>
 at /etc/puppet/environments/webhook/modules/r10k/manifests/webhook/config.pp:68 on node puppet.nelson.va
Warning: Not using cache on failed catalog
Error: Could not retrieve catalog; skipping run
```
If I remove the bang character in webhook.yaml.erb, it proceeds, but generates a `/etc/webhook.yaml` that is not parseable:
```
[root@puppet ~]# puppet agent -t --environment webhook
Info: Retrieving pluginfacts
Info: Retrieving plugin
Info: Loading facts
Info: Caching catalog for puppet.nelson.va
Info: Applying configuration version '1430660173'
Notice: /Stage[main]/R10k::Webhook::Config/File[webhook.yaml]/ensure: defined content as '{md5}2c6a6f93e3debc7e2ec148fc90b71949'
Info: /Stage[main]/R10k::Webhook::Config/File[webhook.yaml]: Scheduling refresh of Service[webhook]
Notice: /File[/etc/sysconfig/iptables]/seluser: seluser changed 'unconfined_u' to 'system_u'
Error: /Stage[main]/R10k::Webhook/Service[webhook]: Failed to call refresh: Could not start Service[webhook]: Execution of '/sbin/service webhook start' returned 1: Starting webhook: /usr/local/bin/webhook:36:in `[]': can't convert String into Integer (TypeError)
        from /usr/local/bin/webhook:36
[FAILED]
Error: /Stage[main]/R10k::Webhook/Service[webhook]: Could not start Service[webhook]: Execution of '/sbin/service webhook start' returned 1: Starting webhook: /usr/local/bin/webhook:36:in `[]': can't convert String into Integer (TypeError)
        from /usr/local/bin/webhook:36
[FAILED]
Notice: /Stage[main]/Profile::Base/Exec[shosts.equiv]/returns: executed successfully
Notice: Finished catalog run in 7.28 seconds

[root@puppet ~]# cat /etc/webhook.yaml
---
  - - port
    - "8088"
  - - use_mcollective
    - false
  - - client_cfg
    - /var/lib/peadmin/.mcollective
  - - prefix
    - false
  - - client_timeout
    - "120"
  - - access_logfile
    - /var/log/webhook/access.log
  - - pass
    - puppet
  - - certpath
    - /var/lib/peadmin/.mcollective.d
  - - discovery_timeout
    - "10"
  - - certname
    - peadmin
  - - prefix_command
    - "/bin/echo example"
  - - use_mco_ruby
    - false
  - - bind_address
    - "0.0.0.0"
  - - enable_ssl
    - true
  - - r10k_deploy_arguments
    - "-pv"
  - - mco_logfile
    - /var/log/webhook/mco_output.log
  - - protected
    - true
  - - user
    - puppet
```
If I fix that manually (to 'username: root', etc) and start the service, it starts appropriately.

FWIW, going back to the oldest version of the yaml erb (https://github.com/mvisonneau/r10k/commit/a86e8ace54f6501dcbf3645174f0d5253f6553d3#diff-9e98459bee65f717707225ca52107dae) generates `/etc/webhook.yaml` in the correct format and skips the keys with an undef value and causes everything to work. I can submit a PR to revert the 4/8/2015 changes, but I'm not sure what the reasoning was to change it in the first place. My ruby-fu is weak, but I'm guessing that in newer ruby versions to_yaml might have different behavior? Let me know if I should submit that PR.

Environment:
zack/r10k: 2.7.3
Puppet: 3.7.5 (FOSS)
Master OS: CentOS 6.6
Master setup: All-in-one, no mcollective (use_mcollective is false)
Ruby: 1.8.7

**Comment by rnelson0** [comment]
https://github.com/rnelson0/r10k-1/commit/a9dda33a64cc2ca7066f054cdd248daa6c849287 is my local fix

**Comment by elyscape** [comment]
This is a result of [the current source for `webhook.yaml.erb`](https://github.com/acidprime/r10k/blob/v2.7.3/templates/webhook.yaml.erb#L1) interacting poorly with Ruby 1.8.7.  To start with, `Hash#select!` does not exist in Ruby 1.8.7.  However, even if we change it to `Hash#select`, Ruby 1.8.7's `Hash#select` method [returns an array of `[key,value]` pairs](http://ruby-doc.org/core-1.8.7/Hash.html#method-i-select) instead of a new hash.  All subsequent versions of Ruby instead return a new hash.  A potential solution would be to change `@webhook_hash.select!{ |k,v| v != :undef }.to_yaml` to `Hash[@webhook_hash.select{ |k,v| v != :undef }].to_yaml`, which would convert the array back to a hash on Ruby 1.8.7 and do nothing on newer versions of Ruby.

That being said, I question the purpose of the call to `Hash#select` in the first place.  The commit message for 9382b10 indicates that this was done to avoid serializing `:undef`, but it doesn't make clear the circumstances under which that would ever happen.  @acidprime, could you expand on that some?  If that's an actual issue that occurs, then it's worth modifying the template to handle that, but if not then perhaps we should revert to `@webhook_hash.to_yaml` for simplicity's sake.

**Comment by acidprime** [comment]
https://github.com/acidprime/r10k/commit/9382b1091422f3c69d191a5601a306c123727c28 was mostly cosmetic I think I for simplicities sake @rnelson0 might be the best "fix"

**Comment by acidprime** [comment]
released in 2.7.4 , thanks all
